node/impl/full/eth.go

```diff
@@ -1270,6 +1307,35 @@ func (e *EthEventHandler) EthGetLogs(ctx context.Context, filterSpec *ethtypes.E
 	return ethFilterResultFromEvents(ctx, ces, e.SubManager.StateAPI)
 }
 
+func (e *EthEventHandler) waitForHeightProcessed(ctx context.Context, height abi.ChainEpoch) error {
+	ei := e.EventFilterManager.EventIndex
+	b, err := ei.IsHeightProcessed(ctx, uint64(height))
```

Comment:
There's a race in here between calling `IsHeightProcessed` and `SubscribeUpdates` where the height may be processed between the two calls. You might need a `SubscribeUpdatesIfHeightProcessed` that can do it atomically for you; then you're going to have to make sure you hold a lock internally between the query and getting the channel in place.

A slight variation on this would be to ditch the "is height processed" check entirely, always subscribe, but immediately send the last processed tipset as soon as you've added the subscription (which also needs to be done while holding the lock) so that you always get a `EventIndexUpdate` with the current head which may be what you want, so it's like "is height processed" but we find out while watching the channel, which should have an event waiting for us.

Reply:
Fixed this by making the event stateless and making all clients defer to the DB as the source of truth after being notified that the DB has been updated.